[Patch] Enhanced Logging, AWS Client Initialization, and Database Handling with Test Mode Support

Kydoimos97 opened the pull request:
**Summary:**
This PR introduces significant improvements to the `WrenchCL` codebase, with a focus on better handling of logging, client initialization, database operations, and exception management. Key changes include:

1. **Logging Enhancements**:
   - Added colored and formatted logging in `WrenchLogger`, including customizable color settings and structured error messages for better readability, especially in tracebacks.
   - Deprecated certain logging aliases while enhancing existing log levels like `context`, `flow`, and `error`.
   - Introduced new logging methods with improved formatting for multi-line and ANSI-style color support.

2. **AWS Client Updates in `AwsClientHub`**:
   - Added additional debug information during S3 and other AWS client initializations to better track profiles and region settings.
   - Enhanced error handling by including profile details in logs for better traceability.
**Example Start up Debugs**
		```java
		INFO    : [R-DTDLYO-0352][2024-11-06 23:10:57] GlobalRouter.py:__init__:17  | Initializing GlobalRouter...
		  DEBUG : [R-DTDLYO-0352] _ConfigurationManager.py:__init__:89  | {'aws_profile': 'prod-ltc', 'region_name': None, 'secret_arn': 'arn...kxj', 'openai_api_key': None, 'ssh_server': None, 'ssh_port': 22, 'ssh_user': None, 'ssh_password': None, 'pem_path': None, 'qa_host_check': 'ce5sivkxtgbs', 'db_batch_size': 10000, 'aws_deployment': False}
		  DEBUG : [R-DTDLYO-0352] AwsClientHub.py:get_secret:316 | Creating session with profile: prod-ltc
		  DEBUG : [R-DTDLYO-0352] AwsClientHub.py:_determine_need_for_tunnel:372 | Running on Production: No SSH tunnel required, continuing.
		  DEBUG : [R-DTDLYO-0352] AwsClientHub.py:_determine_need_for_tunnel:410 | Determined no need for SSH tunnel creation.
		  DEBUG : [R-DTDLYO-0352] AwsClientHub.py:_init_other_client:292 | Creating session with profile: prod-ltc and region: None for service [secretsmanager]
		  DEBUG : [R-DTDLYO-0352] AwsClientHub.py:_init_other_client:294 | secretsmanager client initialized with region: None and profile: prod-ltc
		  DEBUG : [R-DTDLYO-0352] AwsClientHub.py:get_db_uri:128 | Constructed DB URI with endpoint: prod-wrench-ai.c3zncwpdk0m7.us-east-1.rds.amazonaws.com, port: 5432, user: masteruser
		WARNING : [R-DTDLYO-0352][2024-11-06 23:10:58] RdsServiceGateway.py:set_test_mode:59  | Test mode activated, database commits will not be commited.
		  DEBUG : [R-DTDLYO-0352] AwsClientHub.py:_init_s3_client:275 | Creating session with profile: prod-ltc and region: None for service [s3]...
		  DEBUG : [R-DTDLYO-0352] AwsClientHub.py:_init_s3_client:277 | S3 client initialized with region: None
		  DEBUG : [R-DTDLYO-0352] S3ServiceGateway.py:__init__:48  | S3ServiceGateway initialized with S3 client.
		```
3. **Database Handling Improvements in `RdsServiceGateway`**:
   - Introduced a `test_mode` for database operations, allowing transactions to be rolled back instead of committed, supporting safer testing. 
   		- 	Set with enviornment variable: `os.environ["TEST_MODE"] = 'True'`
   - Improved `update_database` to handle a large variety of payload types including `Any, List[Any]`

4. **Custom Exceptions**:
   - Added new custom exceptions in `WrenchCL/Exceptions`, such as `IncompleteInitializationError`, `InitializationError`, `ArgumentTypeError`, and `ArgumentValueError` to streamline error handling and improve clarity in debugging.

5. **Test Updates**:
   - Adjusted test cases in `tests/test_tools.py` to verify enhanced logging capabilities, especially for stack traces and improved trace labels.

Kydoimos97 committed:
Add custom exception classes in WrenchCL/Exceptions

Introduced custom exceptions to handle specific error cases: `InitializationError`, `IncompleteInitializationError`, `ArgumentTypeError`, and `ArgumentValueError`. This improves error handling and makes the codebase more robust.

Kydoimos97 committed:
Merge remote-tracking branch 'origin/v2' into v2

Kydoimos97 committed:
Enhance logging module with color support and simplification

Added color formatting support for log messages and optimized the traceback formatting. Deprecated several specific log levels, simplifying the overall logging interface. Adjusted formatting logic within the `_log` and `_handlerFormat` methods to handle color and level-specific reset.

Kydoimos97 committed:
Add trailing newline to JsonSerializer and MaybeMonad

This commit adds a trailing newline to the JsonSerializer and MaybeMonad modules to comply with coding standards. The change ensures proper formatting and consistency across the codebase.

Kydoimos97 committed:
Fix logger flow test string to match updated exec trace format

The test case for the logger flow was updated to check for the correct execution trace format after recent changes in logging. This ensures that the test accurately reflects the new log output structure and prevents false negatives.

Kydoimos97 committed:
feat: Add test mode, payload flexibility, and aliasing in RdsServiceGateway

- Enhanced `update_database` to support additional payload types, including single values, lists, and DataFrames
- Added debug logging for test mode rollbacks and payload type detection to improve traceability
- Implemented `set_test_mode` method to globally toggle test mode for transaction rollbacks
- Introduced aliases (`insert_data`, `fetch_data`, etc.) for `update_database` and `get_data` for improved readability and usage consistency

Kydoimos97 committed:
Add detailed logging for AWS client session creation

Enhanced debug logging to include profile name and region during AWS client session creation. This helps track and diagnose session initialization with different profiles and regions.

Kydoimos97 committed:
Change log levels and format of logger messages

Updated various logger messages from 'info' to 'warning' for test mode activations and adjusted formatting of query-related logs. This improves clarity and helps identify when the system is running in test mode more easily.

Kydoimos97 committed:
Enhance data logging and traceback formatting in WrenchLogger

Refactor traceback formatting for better readability and update pandas options for improved data display, including handling various display parameters and restoring default settings after logging.

Kydoimos97 committed:
Merge branch 'release' into v2

KevinWrench commented on WrenchCL/Tools/WrenchLogger.py:
```diff
@@ -527,11 +534,17 @@ def error(self, *args: Any, stack_info: Optional[bool] = True, compact: Optional
+            import pandas as pd
+        except ImportError:
+            pd = None
+        option_bu = pd.options
```
AttributeError if pd is None.

Kydoimos97 replied:
You're correct

![image](https://github.com/user-attachments/assets/27f1111e-7030-4a41-8808-34b19c06dda1)


made an update: [`19d325e` (#82)](https://github.com/WrenchAI/WrenchCL/pull/82/commits/19d325e856d4348306207046be03eb94bdc48efc)

that fixes it

KevinWrench commented on WrenchCL/Tools/WrenchLogger.py:
```diff
@@ -527,11 +534,17 @@ def error(self, *args: Any, stack_info: Optional[bool] = True, compact: Optional
         object_name = object_name if object_name else f"Type: {type(data).__name__}"
         formatted_data = self._format_data(data, object_name, content, wrap_length, max_rows, indent=indent)
         self._log_with_color(self.DATA_lvl, formatted_data, Color.BLUE if colorama_imported else None, stack_info, False)
+        pd.options = option_bu
```
The assignment doesn't do anything. Or the one on 543

Kydoimos97 committed:
Handle pandas ImportError with mock class initialization

If pandas is not installed, a mock class for `pd` with `options` attribute is created to avoid AttributeError. This ensures the logger can still function without interrupting the overall flow when pandas is unavailable.

Kydoimos97 committed:
Merge remote-tracking branch 'origin/v2' into v2
